### PR TITLE
Make start of week adjustable when using span

### PR DESCRIPTION
**Pull request by ALee008**
…ng span in combination with frame == "week"

## Pull Request Checklist

Thank you for taking the time to improve Arrow! Before submitting your pull request, please check all *appropriate* boxes:

- [x] 🧪  Added **tests** for changed code.
- [ ] 🛠️  All tests **pass** when run locally (run `tox` or `make test` to find out!).  --> see comments below
- [ ] 🧹  All linting checks **pass** when run locally (run `tox -e lint` or `make lint` to find out!).  --> see comment below
- [x] 📚  Updated **documentation** for changed code.
- [x] ⏩  Code is **up-to-date** with the `master` branch.

If you have *any* questions about your code changes or any of the points above, please submit your questions along with the pull request and we will try our best to help!

## Description of Changes

Closes:  #355 
This is my first PR and I choose this issue because it was labeled as 'good first issue'. I added a new parameter `weekday` that can be used in combination with `frame == "week"`  to choose a start day for `arrow.span()`.

My tests weren't successful. There was an issue with Python3.6 and an error. I ran the tests again against the master code and encountered the same error, see screenshots:

Master Error:
grafik
Master Summary:

grafik

These where the only problems I encountered after implementing my changes.

Furthermore I had problems running the linting, I got the following message:
grafik

I hope we can resolve these issues. Thank you and kind regards.

### CODE REVIEW

**Comment by jadchaar** [comment]
Hi @ALee008, thanks for the PR. I will try to review this in the next couple of weeks, I am pretty swamped at the moment. Hopefully @krisfremen or @systemcatch could review it in the meantime.

**Comment by jadchaar** [comment]
Also, the tests are currently failing because of an XFailed pytest test case that is decreasing coverage. I will submit a patch to get this resolved.

**Comment by ALee008** [comment]
Hey @jadchaar ,
much appriciated. I am looking forward to your feedback.

**Comment by systemcatch** [comment]
Hey @ALee008 have you installed all the dependencies needed to run the tests? You need to install the following;

```shell
pip install tox
pip install pre-commit
pip install mypy
pip install -r requirements.txt
```

Also if you have `tox` already try `tox -e lint`.

**Comment by ALee008** [comment]
> Hey @ALee008 have you installed all the dependencies needed to run the tests? You need to install the following;
> 
> ```shell
> pip install tox
> pip install pre-commit
> pip install mypy
> pip install -r requirements.txt
> ```
> 
> Also if you have `tox` already try `tox -e lint`.

Hey @systemcatch ,

I haven't installed these. Will do and re-test. Thank you.

**Comment by ALee008** [comment]
Hey @systemcatch ,

all linting were successful but ´test_shift_kiritimati´ is still failing. How should I proceed now?

**Comment by jadchaar** [comment]
Try pulling the latest code @ALee008, it seems like the CI runs are only failing on linting now: https://github.com/arrow-py/arrow/pull/934/checks

**Comment by systemcatch** [comment]
@ALee008 what is the output of `pip list`?

**Comment by ALee008** [comment]
> @ALee008 what is the output of `pip list`?

My ´pip list´result (running in an virtual env). 

Package                       Version
----------------------------- ----------
alabaster                     0.7.12
appdirs                       1.4.4
attrs                         20.3.0
Babel                         2.9.0
certifi                       2020.12.5
cfgv                          3.2.0
chardet                       4.0.0
coverage                      5.5
dateparser                    1.0.0
distlib                       0.3.1
docutils                      0.16
filelock                      3.0.12
identify                      2.0.0
idna                          2.10
imagesize                     1.2.0
importlib-metadata            3.7.0
iniconfig                     1.1.1
Jinja2                        2.11.3
MarkupSafe                    1.1.1
mypy                          0.812
mypy-extensions               0.4.3
nodeenv                       1.5.0
packaging                     20.9
pip                           21.0.1
pluggy                        0.13.1
pre-commit                    2.10.1
py                            1.10.0
Pygments                      2.8.0
pyparsing                     2.4.7
pytest                        6.2.2
pytest-cov                    2.11.1
pytest-mock                   3.5.1
python-dateutil               2.8.1
pytz                          2019.3
PyYAML                        5.4.1
regex                         2020.11.13
requests                      2.25.1
setuptools                    54.0.0
simplejson                    3.17.2
six                           1.15.0
snowballstemmer               2.1.0
Sphinx                        3.5.1
sphinx-autodoc-typehints      1.11.1
sphinxcontrib-applehelp       1.0.2
sphinxcontrib-devhelp         1.0.2
sphinxcontrib-htmlhelp        1.0.3
sphinxcontrib-jsmath          1.0.1
sphinxcontrib-qthelp          1.0.3
sphinxcontrib-serializinghtml 1.1.4
toml                          0.10.2
tox                           3.22.0
typed-ast                     1.4.2
typing-extensions             3.7.4.3
tzlocal                       2.1
urllib3                       1.26.3
virtualenv                    20.4.2
zipp                          3.4.0

**Comment by ALee008** [comment]
> Try pulling the latest code @ALee008, it seems like the CI runs are only failing on linting now: https://github.com/arrow-py/arrow/pull/934/checks

Will do and retry.

File: arrow/arrow.py [hunk 1]
```diff
@@ -504,6 +504,7 @@ def span(
         count: int = 1,
         bounds: _BOUNDS = "[)",
         exact: bool = False,
+        weekday: int = 1
```
Comment (systemcatch) [inline arrow/arrow.py:507]: Comma at the end here should fix the lint failure.

**Review by systemcatch — CHANGES_REQUESTED** [review]

File: arrow/arrow.py (hunk 1 above)
Comment (ALee008) [reply arrow/arrow.py]: I added the missing comma. What now?

Comment (systemcatch) [reply arrow/arrow.py]: We look over your code :wink:

File: arrow/arrow.py [hunk 2]
```diff
@@ -517,6 +518,8 @@ def span(
         :param exact: (optional) whether to have the start of the timespan begin exactly
             at the time specified by ``start`` and the end of the timespan truncated
             so as not to extend beyond ``end``.
+        :param weekday: (optional) only in combination with frame == "week". Conform with self.isoweekday(), 1 = Monday,
```
Comment (systemcatch) [inline arrow/arrow.py:521]: I tend to think we should call this `week_start` for clarity.

Comment (systemcatch) [inline arrow/arrow.py:521]:
```suggestion
        :param week_start: (optional) only used in combination with frame == "week". Follows  isoweekday() where Monday is 1 and Sunday is 7
```

File: arrow/arrow.py [hunk 3]
```diff
@@ -537,6 +540,8 @@ def span(
             >>> arrow.utcnow().span('day', bounds='[]')
             (<Arrow [2013-05-09T00:00:00+00:00]>, <Arrow [2013-05-10T00:00:00+00:00]>)
 
+            >>> arrow.utcnow().span('week', weekday=6)
+            (<Arrow [2021-02-20T00:00:00+00:00]>, <Arrow [2021-02-26T23:59:59.999999+00:00]>)
```
Comment (systemcatch) [inline arrow/arrow.py:544]: Maybe compare this with a normal `span("week")` to highlight the difference.

File: tests/test_arrow.py [hunk 4]
```diff
@@ -1754,6 +1754,11 @@ def test_span_week(self):
         assert floor == datetime(2013, 2, 11, tzinfo=tz.tzutc())
         assert ceil == datetime(2013, 2, 17, 23, 59, 59, 999999, tzinfo=tz.tzutc())
 
+        floor, ceil = self.arrow.span("week", weekday=6)
```
Comment (systemcatch) [inline tests/test_arrow.py:1757]: We should also test what happens if a bad value is provided (i.e. `week_start=55`)

**Review by systemcatch — CHANGES_REQUESTED** [review]

File: arrow/arrow.py (hunk 2 above)
Comment (ALee008) [reply arrow/arrow.py]: Hey,
I changed the parameter name to week_start.

File: arrow/arrow.py (hunk 3 above)
Comment (ALee008) [reply arrow/arrow.py]: Hey,
extended doc string in latest commit.

File: arrow/arrow.py [hunk 5]
```diff
@@ -537,6 +540,14 @@ def span(
             >>> arrow.utcnow().span('day', bounds='[]')
             (<Arrow [2013-05-09T00:00:00+00:00]>, <Arrow [2013-05-10T00:00:00+00:00]>)
 
+            >>> arrow.utcnow().span('week')
+            (<Arrow [2021-02-22T00:00:00+00:00]>, <Arrow [2021-02-28T23:59:59.999999+00:00]>)
+
+            >>> arrow.utcnow().span('week', week_start=1)
+            (<Arrow [2021-02-22T00:00:00+00:00]>, <Arrow [2021-02-28T23:59:59.999999+00:00]>)
```
Comment (systemcatch) [inline arrow/arrow.py:547]: No need for these line as it uses the default value above.

File: tests/test_arrow.py [hunk 6]
```diff
@@ -1754,6 +1754,11 @@ def test_span_week(self):
         assert floor == datetime(2013, 2, 11, tzinfo=tz.tzutc())
         assert ceil == datetime(2013, 2, 17, 23, 59, 59, 999999, tzinfo=tz.tzutc())
 
+        floor, ceil = self.arrow.span("week", week_start=6)
+
+        assert floor == datetime(2013, 2, 9, tzinfo=tz.tzutc())
+        assert ceil == datetime(2013, 2, 15, 23, 59, 59, 999999, tzinfo=tz.tzutc())
+
```
Comment (systemcatch) [inline tests/test_arrow.py:1761]: We still need a test that checks what happens when a bad week_start number is passed.

**Review by systemcatch — CHANGES_REQUESTED** [review]

File: tests/test_arrow.py (hunk 6 above)
Comment (ALee008) [reply tests/test_arrow.py:1761]: You're right. There is no check whether the parameter week_day is between 1 and 7. Setting week_day to 55 would return a week in the future. I could raise a `ValueError` if `1 <= week_day <= 7` is not met.

Comment (systemcatch) [reply tests/test_arrow.py:1761]: @Alee008 sounds like a good idea along with a test for that.

**Comment by systemcatch** [comment]
Great, @jadchaar, @krisfremen any further thoughts here?

File: arrow/arrow.py [hunk 7]
```diff
@@ -563,7 +573,11 @@ def span(
             floor = self.__class__(*values, tzinfo=self.tzinfo)  # type: ignore
 
             if frame_absolute == "week":
-                floor = floor.shift(days=-(self.isoweekday() - 1))
+                delta = 0
+                # if weekday is greater than self.isoweekday() go back one week by setting delta = 7
+                if week_start > self.isoweekday():
+                    delta = 7
+                floor = floor.shift(days=-(self.isoweekday() - week_start) - delta)
```
Comment (jadchaar) [inline arrow/arrow.py:580]: Could a modulo be used here rather than a subtraction to wrap around?

Comment (ALee008) [reply arrow/arrow.py:580]: Hey @jadchaar ,
I made attempts using modulo, but without success. My assumption is that provided date must still be in the week that is spaned. Modulo might wrap the week but I need to forceably move back a week, hence -7 days.
Example:
``` # 2021-03-18 is thursday, 5 is isoweekday of Friday```
```arrow.get("2021-03-18").span('week', week_start=5)```
```>>> (<Arrow [2021-03-12T00:00:00+00:00]>, <Arrow [2021-03-18T23:59:59.999999+00:00]>)```

Does this make sense to you?

File: arrow/arrow.py [hunk 8]
```diff
@@ -563,7 +573,11 @@ def span(
             floor = self.__class__(*values, tzinfo=self.tzinfo)  # type: ignore
 
             if frame_absolute == "week":
-                floor = floor.shift(days=-(self.isoweekday() - 1))
+                delta = 0
+                # if week_start is greater than self.isoweekday() go back one week by setting delta = 7
+                if week_start > self.isoweekday():
+                    delta = 7
```
Comment (jadchaar) [inline arrow/arrow.py:579]: This could probably be made into a clean 1 liner: `delta = 7 if week_start > self.isoweekday() else 0` and include the comment on top of it :).

File: arrow/arrow.py [hunk 9]
```diff
@@ -537,7 +540,14 @@ def span(
...
             (<Arrow [2013-05-09T00:00:00+00:00]>, <Arrow [2013-05-10T00:00:00+00:00]>)
 
+            >>> arrow.utcnow().span('week')
+            (<Arrow [2021-02-22T00:00:00+00:00]>, <Arrow [2021-02-28T23:59:59.999999+00:00]>)
+
+            >>> arrow.utcnow().span('week', week_start=6)
+            (<Arrow [2021-02-20T00:00:00+00:00]>, <Arrow [2021-02-26T23:59:59.999999+00:00]>)
         """
+        if not 1 <= week_start <= 7:
+            raise ValueError("condition 1 <= week_start <= 7 not met")
```
Comment (jadchaar) [inline arrow/arrow.py:550]: Let's make this message a bit more readable like "Week start argument must be between 1 and 7."

File: arrow/arrow.py [hunk 10]
```diff
@@ -517,6 +518,8 @@ def span(
         :param exact: (optional) whether to have the start of the timespan begin exactly
             at the time specified by ``start`` and the end of the timespan truncated
             so as not to extend beyond ``end``.
+        :param week_start: (optional) only used in combination with frame == "week". Follows isoweekday() where Monday
```
Comment (jadchaar) [inline arrow/arrow.py:521]: I think we should change the comment a bit to say "only used in combination with the week timeframe. Follows..."

File: arrow/arrow.py (hunk 8 above)
Comment (jadchaar) [reply arrow/arrow.py:579]: Thinking about it though... since we now have the `not 1 <= week_start <= 7` check above, isn't it never going to be the case that week_start is greater than isoweekday?

**Review by jadchaar — CHANGES_REQUESTED** [review]
Almost there @ALee008 thanks for sticking with us. A few comments :)

File: arrow/arrow.py (hunk 8 above)
Comment (ALee008) [reply arrow/arrow.py:579]: `isoweekday` refers to the `arrow` object, which might be 3 (wednesday). Assuming the user wants to span from last Thursday, `week_start` must be set to 4. This would span the week from last Thursday to coming Thursday.

File: arrow/arrow.py (hunk 9 above)
Comment (ALee008) [reply arrow/arrow.py:550]: Changed text to "week_start argument must be between 1 and 7."

File: arrow/arrow.py (hunk 10 above)
Comment (ALee008) [reply arrow/arrow.py:521]: Updated comment.

File: arrow/arrow.py (hunk 8 above)
Comment (ALee008) [reply arrow/arrow.py:579]: Updated the code and made a one liner.

**Comment by ALee008** [comment]
Hey,
lint docs test failed. But I can't interpret the error message:
```[...]
ERROR: InvocationError for command /usr/bin/make html 'SPHINXOPTS=-W --keep-going' (exited with code 2)
___________________________________ summary ____________________________________
ERROR:   docs: commands failed
Error: Process completed with exit code 1.
```

File: arrow/arrow.py [hunk 11]
```diff
@@ -537,7 +540,14 @@ def span(
             >>> arrow.utcnow().span('day', bounds='[]')
             (<Arrow [2013-05-09T00:00:00+00:00]>, <Arrow [2013-05-10T00:00:00+00:00]>)
 
+            >>> arrow.utcnow().span('week')
+            (<Arrow [2021-02-22T00:00:00+00:00]>, <Arrow [2021-02-28T23:59:59.999999+00:00]>)
+
+            >>> arrow.utcnow().span('week', week_start=6)
+            (<Arrow [2021-02-20T00:00:00+00:00]>, <Arrow [2021-02-26T23:59:59.999999+00:00]>)
         """
```
Comment (systemcatch) [inline arrow/arrow.py:548]: @ALee008 I think we need a blank line at the end of the docstring here.

```suggestion
            (<Arrow [2021-02-20T00:00:00+00:00]>, <Arrow [2021-02-26T23:59:59.999999+00:00]>)
            
        """
```

**Comment by ALee008** [comment]
> Hey,
> lint docs test failed. But I can't interpret the error message:
> 
> ```
> ERROR: InvocationError for command /usr/bin/make html 'SPHINXOPTS=-W --keep-going' (exited with code 2)
> ___________________________________ summary ____________________________________
> ERROR:   docs: commands failed
> Error: Process completed with exit code 1.
> ```

It didn't seem to fix the problem. Any suggestions?

**Comment by systemcatch** [comment]
Weird I don't understand why this is still failing, @jadchaar any ideas?

**Comment by jadchaar** [comment]
Yeah I fixed it--it was a trailing whitespace.

**Comment by jadchaar** [comment]
Hmm it seems like Github is rejecting my commit to this branch. I got an error about not having permission, so it seems to not be associated with a branch:
Screen Shot 2021-03-22 at 3 24 48 PM

@ALee008 mind just removing that trailing whitespace on the line I changed in 4c1306a and pushing it?

**Comment by ALee008** [comment]
> Hmm it seems like Github is rejecting my commit to this branch. I got an error about not having permission, so it seems to not be associated with a branch:
> Screen Shot 2021-03-22 at 3 24 48 PM
> 
> @ALee008 mind just removing that trailing whitespace on the line I changed in [4c1306a](https://github.com/arrow-py/arrow/commit/4c1306acd24facde15f5aa5eba248f639fbdd6c9) and pushing it?

Will do an push.

**Comment by systemcatch** [comment]
@jadchaar can you take a look at this again? I'm afk until Saturday night but will try to fix this then otherwise.

**Comment by jadchaar** [comment]
> @jadchaar can you take a look at this again? I'm afk until Saturday night but will try to fix this then otherwise.

Yup fixed, it was a docs indent issue.

**Comment by ALee008** [comment]
> > @jadchaar can you take a look at this again? I'm afk until Saturday night but will try to fix this then otherwise.
> 
> Yup fixed, it was a docs indent issue.

Hey @jadchaar,
what are the next steps? Do I need to do something now?

**Comment by jadchaar** [comment]
> > > @jadchaar can you take a look at this again? I'm afk until Saturday night but will try to fix this then otherwise.
> > 
> > 
> > Yup fixed, it was a docs indent issue.
> 
> Hey @jadchaar,
> what are the next steps? Do I need to do something now?

Hi @ALee008, apologies. I am still trying to understand the logic of finding the delta: `delta = 7 if week_start > self.isoweekday() else 0`. Chris brought up a good point with me: if week_start is 4 and isoweekday is 2, don't we need to go back 9 days? Also, it would be nice to expand the tests to cover more cases with more week_starts and more edge cases.

**Comment by jadchaar** [comment]
Mind adding a few more test cases that cover both the cases you mention @ALee008? Then we can work to finally get this merged :).

**Review by systemcatch — APPROVED** [review]
Thanks for your perseverance @ALee008, we're good to merge this in now.

**Review by jadchaar — APPROVED** [review]
LGTM

**Comment by ALee008** [comment]
> Thanks for your perseverance @ALee008, we're good to merge this in now.

I also have to thank you. You guys are doing a great job!

**Comment by jadchaar** [comment]
> > Thanks for your perseverance @ALee008, we're good to merge this in now.
> 
> I also have to thank you. You guys are doing a great job!

Thanks @ALee008 we appreciate it! We are trying to keep Arrow awesome :), but we cannot forget how amazing the community has been in helping us with that mission!
